### PR TITLE
Allow using outgoing channel id for onion messages

### PR DESCRIPTION
**Pull request by thomash-acinq**
Allow using a short channel id instead of a public key to designate the next node a message should be relayed to.

### CODE REVIEW

File: eclair-core/src/test/scala/fr/acinq/eclair/integration/basic/payment/OfferPaymentSpec.scala [hunk 1]
```diff
@@ -138,8 +139,9 @@ class OfferPaymentSpec extends FixtureSpec with IntegrationPatience {
     val recipientKey = randomKey()
     val pathId = randomBytes32()
     val offerPaths = routes.map(route => {
-      route.nodes.dropRight(1).map(IntermediateNode(_))
-      buildRoute(randomKey(), route.nodes.dropRight(1).map(IntermediateNode(_)), Recipient(route.nodes.last, Some(pathId)))
+      val ourNode = route.nodes.last
+      val intermediateNode = route.nodes.dropRight(1).map(IntermediateNode(_)) ++ route.dummyHops.map(_ => IntermediateNode(ourNode))
+      buildRoute(randomKey(), intermediateNode, Recipient(route.nodes.last, Some(pathId)))
```
Comment (t-bast) [inline eclair-core/src/test/scala/fr/acinq/eclair/integration/basic/payment/OfferPaymentSpec.scala:144]: nit:
```suggestion
      val ourNodeId = route.nodes.last
      val intermediateNodes = route.nodes.dropRight(1).map(IntermediateNode(_)) ++ route.dummyHops.map(_ => IntermediateNode(ourNodeId))
      buildRoute(randomKey(), intermediateNodes, Recipient(ourNodeId, Some(pathId)))
```

File: eclair-core/src/test/scala/fr/acinq/eclair/integration/basic/payment/OfferPaymentSpec.scala [hunk 2]
```diff
@@ -350,4 +352,38 @@ class OfferPaymentSpec extends FixtureSpec with IntegrationPatience {
...
+    val receivingRoute = ReceivingRoute(Seq(bob.nodeId, carol.nodeId), maxFinalExpiryDelta)
+
+    val handler = carol.system.spawnAnonymous(offerHandler(amount, Seq(receivingRoute)))
+    carol.offerManager ! OfferManager.RegisterOffer(compactOffer, recipientKey, Some(pathId), handler)
+    val offerPayment = alice.system.spawnAnonymous(OfferPayment(alice.nodeParams, alice.postman, alice.paymentInitiator))
+    val sendPaymentConfig = OfferPayment.SendPaymentConfig(None, connectDirectly = false, maxAttempts = 1, alice.routeParams, blocking = true)
+    offerPayment ! OfferPayment.PayOffer(probe.ref, compactOffer, amount, 1, sendPaymentConfig)
+
+    val payment = verifyPaymentSuccess(compactOffer, amount, probe.expectMsgType[PaymentEvent])
+    assert(payment.parts.length == 1)
```
Comment (t-bast) [inline eclair-core/src/test/scala/fr/acinq/eclair/integration/basic/payment/OfferPaymentSpec.scala:387]: I think that using a `RouteRequest` is confusing, we can do that more easily but just getting explicitly the scid for one of the Bob->Carol channels?

```suggestion
    import f._

    val probe = TestProbe()
    val amount = 25_000_000 msat
    val recipientKey = randomKey()
    val pathId = randomBytes32()

    val blindedRoute = buildRoute(randomKey(), Seq(IntermediateNode(bob.nodeId), IntermediateNode(carol.nodeId)), Recipient(carol.nodeId, Some(pathId)))
    val offer = Offer(None, "test", recipientKey.publicKey, Features.empty, carol.nodeParams.chainHash, additionalTlvs = Set(OfferPaths(Seq(blindedRoute))))
    val scid_bc = getPeerChannels(bob, carol.nodeId).head.data.asInstanceOf[DATA_NORMAL].shortIds.real.toOption.get
    val compactBlindedRoute = buildRoute(randomKey(), Seq(IntermediateNode(bob.nodeId, Some(scid_bc)), IntermediateNode(carol.nodeId, Some(ShortChannelId.toSelf))), Recipient(carol.nodeId, Some(pathId)))
    val compactOffer = Offer(None, "test", recipientKey.publicKey, Features.empty, carol.nodeParams.chainHash, additionalTlvs = Set(OfferPaths(Seq(compactBlindedRoute))))
    assert(compactOffer.toString.length < offer.toString.length)

    val receivingRoute = ReceivingRoute(Seq(bob.nodeId, carol.nodeId), maxFinalExpiryDelta)
    val handler = carol.system.spawnAnonymous(offerHandler(amount, Seq(receivingRoute)))
    carol.offerManager ! OfferManager.RegisterOffer(compactOffer, recipientKey, Some(pathId), handler)
    val offerPayment = alice.system.spawnAnonymous(OfferPayment(alice.nodeParams, alice.postman, alice.paymentInitiator))
    val sendPaymentConfig = OfferPayment.SendPaymentConfig(None, connectDirectly = false, maxAttempts = 1, alice.routeParams, blocking = true)
    offerPayment ! OfferPayment.PayOffer(probe.ref, compactOffer, amount, 1, sendPaymentConfig)
    val payment = verifyPaymentSuccess(compactOffer, amount, probe.expectMsgType[PaymentEvent])
    assert(payment.parts.length == 1)
```

File: eclair-core/src/test/scala/fr/acinq/eclair/io/MessageRelaySpec.scala [hunk 3]
```diff
@@ -40,16 +40,17 @@ class MessageRelaySpec extends ScalaTestWithActorTestKit(ConfigFactory.load("app
   val aliceId: PublicKey = Alice.nodeParams.nodeId
   val bobId: PublicKey = Bob.nodeParams.nodeId
 
-  case class FixtureParam(relay: ActorRef[Command], switchboard: TestProbe, peerConnection: TypedProbe[Nothing], peer: TypedProbe[Peer.RelayOnionMessage], probe: TypedProbe[Status])
+  case class FixtureParam(relay: ActorRef[Command], switchboard: TestProbe, register: TestProbe, peerConnection: TypedProbe[Nothing], peer: TypedProbe[Peer.RelayOnionMessage], probe: TypedProbe[Status])
```
Comment (t-bast) [inline eclair-core/src/test/scala/fr/acinq/eclair/io/MessageRelaySpec.scala:43]: Can you add a test that uses the scid?

File: eclair-core/src/test/scala/fr/acinq/eclair/message/PostmanSpec.scala [hunk 4]
```diff
@@ -96,8 +102,12 @@ class PostmanSpec extends ScalaTestWithActorTestKit(ConfigFactory.load("applicat
     assert(target == recipientKey.publicKey)
     waitingForRoute ! MessageRoute(Seq.empty, target)
 
-    val RelayMessage(messageId, _, nextNodeId, _, _, Some(replyTo)) = switchboard.expectMessageType[RelayMessage]
+    val Peer.Connect(nextNodeId, _, replyTo1, _) = switchboard.expectMessageType[Peer.Connect]
     assert(nextNodeId == recipientKey.publicKey)
+    val peerConnection = TestProbe[Any]()
+    val peer = TestProbe[Any]()
+    replyTo1 ! PeerConnection.ConnectionResult.AlreadyConnected(peerConnection.ref.toClassic, peer.ref.toClassic)
+    val Peer.RelayOnionMessage(messageId, _, Some(replyTo)) = peer.expectMessageType[Peer.RelayOnionMessage]
```
Comment (t-bast) [inline eclair-core/src/test/scala/fr/acinq/eclair/message/PostmanSpec.scala:110]: This is worth extracting to a helper function instead of duplicating it.

**Review by t-bast — COMMENTED** [review]
Looks mostly good to me :+1:

File: eclair-core/src/test/scala/fr/acinq/eclair/message/PostmanSpec.scala (hunk 4 above)
Comment (thomash-acinq) [reply eclair-core/src/test/scala/fr/acinq/eclair/message/PostmanSpec.scala]: Done

File: eclair-core/src/test/scala/fr/acinq/eclair/io/MessageRelaySpec.scala (hunk 3 above)
Comment (thomash-acinq) [reply eclair-core/src/test/scala/fr/acinq/eclair/io/MessageRelaySpec.scala:44]: Done

**Review by t-bast — APPROVED** [review]

**Pull request merged**